tiatoolbox/models/dataset/abc.py

```diff
@@ -0,0 +1,152 @@
+# ***** BEGIN GPL LICENSE BLOCK *****
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software Foundation,
+# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+#
+# The Original Code is Copyright (C) 2021, TIALab, University of Warwick
+# All rights reserved.
+# ***** END GPL LICENSE BLOCK *****
+
+
+import torch
+import pathlib
+import numpy as np
+import os
+
+from tiatoolbox.utils.misc import imread
+
+
+class ABCPatchDataset(torch.utils.data.Dataset):
+    """Defines abstract base class for patch dataset.
+
+    Attributes:
+        return_labels (bool, False): `__getitem__` will return both the img and
+        its label. If `label_list` is `None`, `None` is returned
+
+        preproc_func: Preprocessing function used to transform the input data. If
+        supplied, then torch.Compose will be used on the input preproc_list.
+        preproc_list is a list of torchvision transforms for preprocessing the image.
+        The transforms will be applied in the order that they are given in the list.
+        https://pytorch.org/vision/stable/transforms.html.
+
+    """
+
+    def __init__(self, preproc_func=None):
+        super().__init__()
+        self.set_preproc_func(preproc_func)
+        self.data_is_npy_alike = False
+        self.input_list = []
+        self.label_list = []
+
+    def check_input_integrity(self, mode):
+        """Perform check on the input to make sure it is valid."""
```

Comment:
it is valid image or array?
please can you explain in the docs how you are checking validity?